inc/packages/namespace.php

```diff
@@ -0,0 +1,310 @@
+const DID_CACHE_LIFETIME = 5 * MINUTE_IN_SECONDS;
+
+// phpcs:disable WordPress.NamingConventions.ValidVariableName
+
+/**
+ * Bootstrap.
+ *
+ * @return void
+ */
+function bootstrap() {
+	Admin\bootstrap();
+}
+
+/**
+ * Parse DID.
+ *
+ * @param string $id DID.
+ * @return DID|WP_Error
+ */
+function parse_did( string $id ) {
+	if ( ! str_starts_with( $id, 'did:' ) ) {
+		return new WP_Error( 'fair.packages.validate_did.not_did', __( 'ID is not a valid DID.', 'fair' ) );
+	}
+
+	$parts = explode( ':', $id, 3 );
+	if ( count( $parts ) !== 3 ) {
+		return new WP_Error( 'fair.packages.validate_did.not_uri', __( 'DID could not be parsed as a URI.', 'fair' ) );
+	}
+
+	switch ( $parts[1] ) {
+		case PLC::TYPE:
+			return new PLC( $id );
+
+		case Web::TYPE:
+			return new Web( $id );
+
+		default:
+			return new WP_Error( 'fair.packages.validate_id.invalid_type', __( 'Unsupported DID type.', 'fair' ) );
+	}
+}
+
+/**
+ * Return hash of DID.
+ *
+ * This ensures a consistent representation of a DID
+ * regardless of its method, method-specific-id format, or length.
+ *
+ * @param  string $id DID
+ *
+ * @return string|WP_Error
+ */
+function get_did_hash( string $id ) {
+	$did = parse_did( $id );
+	if ( is_wp_error( $did ) ) {
+		return $did;
+	}
+
+	return substr( hash( 'sha256', $did->get_id() ), 0, 6 );
+}
+
+/**
+ * Get DID document.
+ *
+ * @param string $id DID.
+ * @return DIDDocument|WP_Error
+ */
+function get_did_document( string $id ) {
+	$cached = wp_cache_get( $id, 'fair_did_documents', false, $found );
+	if ( $found ) {
+		return $cached;
+	}
+
+	// Parse the DID, then fetch the details.
+	$did = parse_did( $id );
+	if ( is_wp_error( $did ) ) {
+		return $did;
+	}
+
+	$document = $did->fetch_document();
+	if ( is_wp_error( $document ) ) {
+		return $document;
+	}
+
+	wp_cache_set( $id, $document, 'fair_did_documents', DID_CACHE_LIFETIME );
+	return $document;
+}
+
+/**
+ * Fetch metadata for a package.
+ *
+ * @param string $id DID of the package to fetch metadata for.
+ * @return MetadataDocument|WP_Error Metadata document on success, WP_Error on failure.
+ */
+function fetch_package_metadata( string $id ) {
+	$document = get_did_document( $id );
+	if ( is_wp_error( $document ) ) {
+		return $document;
+	}
+
+	// Fetch data from the repository.
+	$service = $document->get_service( SERVICE_ID );
+	if ( empty( $service ) ) {
+		return new WP_Error( 'fair.packages.fetch_metadata.no_service', __( 'DID is not a valid package to fetch metadata for.', 'fair' ) );
+	}
+	$repo_url = $service->serviceEndpoint;
+
+	return fetch_metadata_doc( $repo_url );
+}
+
+/**
+ * Install a plugin from a FAIR DID.
+ *
+ * @param string $id DID of the package to install.
+ * @param string|null $version Version to install. If null, the latest version is installed.
+ * @param WP_Upgrader_Skin $skin Plugin Installer Skin.
+ * @return bool|WP_Error True on success, WP_Error on failure.
+ */
+function install_plugin( string $id, ?string $version = null, $skin ) {
```

Comment:
I'm seeing this on PHP 8.3

```
PHP Deprecated:  Optional parameter $version declared before required parameter $skin is implicitly treated as a required parameter in /Users/afragen/Documents/github/fair-plugin/inc/packages/namespace.php on line 133
```
 FWIW this is locally.

Reply:
Or just ensure that when we call `install_plugin()` we ensure a `$version` parameter.